### PR TITLE
Add max_multipart_parts setting to S3 repository

### PR DESCRIPTION
**Pull request by mhl-b**
Introduce new S3 repository setting - `max_multipart_parts`. Currently, with default settings, to upload large files to S3 we split them into 5TB chunks and then each chunk into parts with up to 100MB size. That means we need to send up to 50k parts, where [S3 limit is 10k](https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html). Attempts to upload files larger than 1TB would fail using default settings.

With new setting we split files by smallest of `chunk_size` or `buffer_size x max_multipart_parts`. With default values of `buffer_size=100MB` and `max_multipart_parts=10_000` maximum "actual" chunk size  will be ~1TB, not 5TB.

This PR is a successor to https://github.com/elastic/elasticsearch/pull/112708

### CODE REVIEW

**Review by nicktindall — APPROVED** [review]
LGTM

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I'll take a deeper look later but I'd definitely like a different name for this setting. Could we call it `max_multipart_parts`?

File: docs/reference/snapshot-restore/repository-s3.asciidoc [hunk 1]
```diff
@@ -262,7 +262,8 @@ multiple deployments may share the same bucket.
 `chunk_size`::
 
     (<<byte-units,byte value>>) Big files can be broken down into chunks during snapshotting if needed.
-    Specify the chunk size as a value and unit, for example:
+    When large file split into chunks, the chunk size will be defined by smallest of `chunk_size`
+    or `buffer_size * max_multipart_parts`. Specify the chunk size as a value and unit, for example:
     `1TB`, `1GB`, `10MB`. Defaults to the maximum size of a blob in the S3 which is `5TB`.
```
Comment (DaveCTurner) [inline docs/reference/snapshot-restore/repository-s3.asciidoc:267]: I never really liked the wording here to start with :) How about something like this? 

```
    (<<byte-units,byte value>>) The maximum size of object that {es} will write to the repository when creating a snapshot. Files which are
    larger than `chunk_size` will be chunked into several smaller objects. {es} may also split a file across multiple objects to satisfy
    other constraints such as the `max_multipart_parts` limit. Defaults to `5TB` which is the
    https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html[maximum size of an object in AWS S3].
```

File: docs/reference/snapshot-restore/repository-s3.asciidoc [hunk 2]
```diff
@@ -292,6 +293,14 @@ include::repository-shared-settings.asciidoc[]
     size allowed by S3. Defaults to `100mb` or `5%` of JVM heap, whichever is
     smaller.
 
+`max_multipart_parts` ::
+
+    (<<number,numeric>>) Maximum number of parts for multipart upload. When large file split into
+    chunks, the chunk size will be defined by smallest of `chunk_size` or `buffer_size * max_multipart_parts`.
+    Default value is 10,000, also see https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html[S3 multipart upload limits].
+    For example, with `buffer_size=100MB` and `max_multipart_parts=10,000` summation of all parts is about 1TB.
+    If chunk_size is set to 5TB then smallest between two would be 1TB.
```
Comment (DaveCTurner) [inline docs/reference/snapshot-restore/repository-s3.asciidoc:302]: I'd be inclined not to dwell too much on the calculations here, how about something like this?

```suggestion
    (<<number,integer>>) The maximum number of parts that {es} will write during a multipart upload of a single object. Files which are
    larger than `buffer_size × max_multipart_parts` will be chunked into several smaller objects. {es} may also split a file across multiple
    objects to satisfy other constraints such as the `chunk_size` limit. Defaults to `10000` which is the
    https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html[maximum number of parts in a multipart upload in AWS S3].
```

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 3]
```diff
@@ -302,6 +309,20 @@ private static Map<String, String> buildLocation(RepositoryMetadata metadata) {
...
+    /**
+     * Calculates S3 object size limit based on 2 constraints: maximum object(chunk) size
+     * and maximum number of parts for multipart upload.
+     * https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
+     *
+     * @param objectSize object size limit in s3 or chunk_size
+     * @param partSize part size in s3 or buffer_size
+     * @param partsNum number of parts(buffers)
+     */
+    static ByteSizeValue objectSizeLimit(ByteSizeValue objectSize, ByteSizeValue partSize, int partsNum) {
```
Comment (DaveCTurner) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java:321]: Nit: could we align the argument names with the names of the variables passed in? I know they're kinda weird names but we can't change the settings' names for legacy reasons and we're following that naming elsewhere.

**Review by DaveCTurner — APPROVED** [review]
LGTM except some suggestions for the docs and a naming nit. We can iterate on the docs separately if you'd prefer.

File: docs/reference/snapshot-restore/repository-s3.asciidoc (hunk 1 above)
Comment (mhl-b) [reply docs/reference/snapshot-restore/repository-s3.asciidoc:267]: + on doc changes, thank you

**Comment by DaveCTurner** [comment]
As discussed [elsewhere](https://github.com/elastic/elasticsearch/pull/114093#issuecomment-2394054801), it'd be good to backport this to `8.x` too. I'm adding the appropriate labels.

**Pull request merged**